kep/42-podgroup-coscheduling/README.md

```diff
@@ -135,12 +141,75 @@ cluster upgrade.
+in coscheduling's core package:
+
+- when podGroup created, we'll create a `queuedPodGroup` in the cache
+- when podGroup updated, we'll do nothing right now
+- when podGroup deleted, we'll clean up the objet in the cache
+
+The structure looks like below:
+
+```golang
+type PodGroupManager struct {
+ // ...
+ // <new>
+ // key is the podGroup name, value is the queued podGroup info.
+ podGroups map[string]*queuedPodGroup
+}
+
+// queuedPodGroup's lifecycle is synchronized with the podGroup object.
+type queuedPodGroup struct {
+ // Timestamp is the podGroup's queued time.
+ // - timestamp will be initialized when the first pod enqueues
+ // - timestamp will be renewed when a pod re-enqueues for failing the scheduling
+ timestamp time.Time
+
+ // status indicated the podGroup's condition, it can be:
+ // - queueing: when the first pod starts scheduling
+ // - queueingFailed: when a pod fails the scheduling
+ // - queueingSucceeded: when the podGroup succeeds in scheduling
+ status string
+}
+```
+
+The workflow is (only works for pods have podGroup):
+
+1. When a new podGroup created, we'll create a queuedPodGroup in the cache.
+2. When the first pod of the podGroup creates, in queueing specifically, we'll initialize the timestamp of the queuedPodGroup,
+also set the status=queueing. This will happen in `Less()` function, and because we'll not change the timestamp, so pods belong
+to the same podGroup will be placed together as a unit.
+3. Schedule the pods of the podGroup one by one, if anyone fails in the scheduling cycle, we'll set the status=queueingFailed in PostFilter,
+and the podGroup will enter into the backoff queue if configured.
+4. When the failed Pod re-enqueues, we'll renew the queuedPodGroup timestamp based on the `queueingFailed` status and also set the status to `queueing`, the new timestamp should be `now+backoffTime` to avoid breaking the podGroup who starts scheduling
+during the backoff time.
+5. When backoff time passes, pods will be re-scheduled.
+6. If podGroup permitted successfully, we'll set the status=queueingSucceeded, and frozen the timestamp.
+7. When the podGroup deleted, we'll remove it from the cache as well.
+
+We refused to initialize the queuedPodGroup's timestamp in creation is because podGroup can be reused, so the creating time could be
+really old but will be queued in the front, that's not what we expect.
+
+One risk here is frozen the timestamp in Step6 sounds a little aggressive, but what we want to do here is trying to make the pods
+behaving as a unit, let's say we have successfully scheduled a podGroup, and we're scheduling another one, then a pod from
+the former podGroup was rebuilt, instead of waiting for the latter podGroup scheduling first, we hope the pod recover in prior,
+although this may lead to the latter podGroup failed in scheduling.
+
+Another risk is we'll initialize or renew the timestamp in the `Less()` function, which will lead to the performance degradation in queueing,
```

Comment:
it's a big no. We should not mutate timestamp in a function that relies on timestamp to do sorting. It'd cause unstable sorting result and make the sorting behavior unpredictable and hard to debug.